instrumentation/github.com/go-kit/kit/otelkit/endpoint.go

```diff
@@ -0,0 +1,135 @@
+
+package otelkit
+
+import (
+	"context"
+	"strconv"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/label"
+	"go.opentelemetry.io/otel/trace"
+
+	otelcontrib "go.opentelemetry.io/contrib"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/sd/lb"
+)
+
+const (
+	tracerName = "go.opentelemetry.io/contrib/instrumentation/github.com/go-kit/kit/otelkit"
+
+	// defaultSpanName is the default endpoint span name to use.
+	defaultSpanName = "gokit/endpoint"
+)
+
+// EndpointMiddleware returns an Endpoint middleware, tracing a Go kit endpoint.
+// This endpoint tracer should be used in combination with a Go kit Transport
+// tracing middleware, generic OpenCensus transport middleware or custom before
+// and after transport functions as service propagation of SpanContext is not
+// provided in this middleware.
+func EndpointMiddleware(options ...EndpointOption) endpoint.Middleware {
+	cfg := &endpointOptions{}
+
+	for _, o := range options {
+		o(cfg)
+	}
+
+	if cfg.TracerProvider == nil {
+		cfg.TracerProvider = otel.GetTracerProvider()
+	}
+
+	tracer := cfg.TracerProvider.Tracer(
+		tracerName,
+		trace.WithInstrumentationVersion(otelcontrib.SemVersion()),
+	)
+
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			operation := cfg.Operation
+			if cfg.GetOperation != nil {
+				if newOperation := cfg.GetOperation(ctx, operation); newOperation != "" {
+					operation = newOperation
+				}
+			}
+
+			spanName := operation
+			if spanName == "" {
+				spanName = defaultSpanName
+			}
+
+			opts := []trace.SpanOption{
+				trace.WithAttributes(cfg.Attributes...),
+				trace.WithSpanKind(trace.SpanKindServer),
+			}
+
+			if cfg.GetAttributes != nil {
+				opts = append(opts, trace.WithAttributes(cfg.GetAttributes(ctx)...))
+			}
+
+			ctx, span := tracer.Start(ctx, spanName, opts...)
+			defer span.End()
+
+			defer func() {
+				if err != nil {
+					if lberr, ok := err.(lb.RetryError); ok {
+						// handle errors originating from lb.Retry
+						for idx, rawErr := range lberr.RawErrors {
+							span.AddEvent(
+								"error",
+								trace.WithAttributes(
+									label.String("error.type", "gokit.lb.retry"),
+									label.String("gokit.lb.retry.count", strconv.Itoa(idx+1)),
+									label.String("error.message", rawErr.Error()),
+								),
+							)
+						}
+
+						span.RecordError(lberr.Final)
+
+						return
+					}
+
+					// generic error
+					span.RecordError(err)
+
+					return
+				}
+
+				// test for business error
+				// business errors are often  successful requests carrying
+				// a business failure that the client can act upon
+				if res, ok := response.(endpoint.Failer); ok && res.Failed() != nil {
+					span.RecordError(res.Failed())
+
+					if cfg.IgnoreBusinessError {
+						span.SetStatus(codes.Unset, "")
+					}
```

Comment:
If the `IgnoreBusinessError` option is `true`, is it intended to still add error event via `span.RecordError` and to just "reset" the status to `Unset`? Or should the span's events also pretend as if there was no error? I was just wondering, as the option description did not make it 100% clear to me.

Reply:
Yeah, that's a little bit confusing concept and is the result of using the original OpenCensus integration as an example.

The idea is to add the error to the span, but not mark the span as a failed one. This in theory will make the request itself successful when carrying an error only meaningful to the user (eg. a validation error), but doesn't indicate an error in the system. Chances are an HTTP middleware will still mark the request as a failed one (I guess), so I'm not sure it's worth keeping.

There is a comment about this a few lines above:

```
				// business errors are often successful requests carrying
				// a business failure that the client can act upon
				// and therefore do not count as failed requests
```

I added another line. Let me know what you think.